msal/imds.py

```diff
@@ -0,0 +1,243 @@
+        return {
+            "error": "invalid_scope",  # Empirically, wrong resource ends up with a vague statusCode=500
+            "error_description": "{}, {}".format(
+                payload.get("statusCode"), payload.get("message")),
+            }
+    except ValueError:
+        logger.debug("IMDS emits unexpected payload: %s", resp.text)
+        raise
+
+
+def _obtain_token_on_service_fabric(
+    http_client, endpoint, identity_header, server_thumbprint, resource,
+):
+    """Obtains token for
+    `Service Fabric <https://learn.microsoft.com/en-us/azure/service-fabric/>`_
+    """
+    # Deployment https://learn.microsoft.com/en-us/azure/service-fabric/service-fabric-get-started-containers-linux
+    # See also https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/identity/azure-identity/tests/managed-identity-live/service-fabric/service_fabric.md
+    # Protocol https://learn.microsoft.com/en-us/azure/service-fabric/how-to-managed-identity-service-fabric-app-code#acquiring-an-access-token-using-rest-api
+    logger.debug("Obtaining token via managed identity on Azure Service Fabric")
+    resp = http_client.get(
+        endpoint,
+        params={"api-version": "2019-07-01-preview", "resource": resource},
+        headers={"Secret": identity_header},
+        )
+    try:
+        payload = json.loads(resp.text)
+        if payload.get("access_token") and payload.get("expires_on"):
+            return {  # Normalizing the payload into OAuth2 format
+                "access_token": payload["access_token"],
+                "expires_in": payload["expires_on"] - int(time.time()),
+                "resource": payload.get("resource"),
+                "token_type": payload["token_type"],
+                }
+        return {
+            "error": "invalid_scope",
+            "error_description": "{}, {}".format(
+                payload.get("statusCode"), payload.get("message")),
+            }
+    except ValueError:
+        logger.debug("IMDS emits unexpected payload: %s", resp.text)
+        raise
+
+
+
+class ManagedIdentity(object):
+    _instance, _tenant = socket.getfqdn(), "managed_identity"  # Placeholders
+
+    def __init__(self, http_client,
+        client_id=None, object_id=None, mi_res_id=None,
+        token_cache=None,
+    ):
+        """Create a managed identity object.
+
+        :param http_client:
+            An http client object. For example, you can use `requests.Session()`.
+
+        :param str client_id:
+            Optional.
+            It accepts the Client ID (NOT the Object ID) of your user-assigned managed identity.
+            If it is None, it means to use a system-assigned managed identity.
+
+        :param token_cache:
+            Optional. It accepts a :class:`msal.TokenCache` instance to store tokens.
+        """
+        if len(list(filter(bool, [client_id, object_id, mi_res_id]))) > 1:
+            raise ValueError("You can use up to one of these: client_id, object_id, mi_res_id")
```

Comment:
```suggestion
            raise ValueError("You can use only one of these: client_id, object_id, mi_res_id")
```

Reply:
I am not an English teacher, but the word "only" seems to have an ambiguity of "choose one and only one". In this context, we mean to say "choose zero or one".

If you do not like "up to", how about I change it to `raise ValueError("You cannot use more than one of these: ...")`?

Reply:
You made a very good point. both `up to`, `cannot use more than` feel a little bit cumbersome to me. `zero or one` is more direct. Perhaps `You can omit id, or pass one of client_id, object_id, mi_res_id`? @dbradish-microsoft is our expert. 😊

Reply:
Sure. Any suggestion is welcome.

@jiasli , for the sake of completeness, we could go with a totally different design here. The little disadvantage is that the calling pattern would be a little bit verbose. But perhaps worth it?

```python
class ManagedIdentityId(object):
    def __init__(self, id):
        self._id = id

class ClientId(ManagedIdentityId):
    pass

class ObjectId(ManagedIdentityId):
    pass

class ResourceId(ManagedIdentityId):
    pass

class ManagedIdentity(object):
    def __init__(self, managed_identity_id, ...):
        if not isinstance(managed_identity_id, (ClientId, ObjectId, ResourceId)):
            raise ValueError("You should provide one of the three objects: ...")
        ...

mi = ManagedIdentity(ClientId("foo"), ...)
```

Reply:
With the latest API change, which is now documented in this PR's description, we can close this topic because the new API simply will not accept multiple IDs. Resolving this conversation.